README.md

```diff
@@ -91,16 +219,627 @@ impl DeserializeFromValue<MyError> for Name {
+```
+
+#### `error`
+
+Customize the error type that can be returned when deserializing this structure
+instead of keeping it generic.
+
+```rust
+use deserr::{DeserializeFromValue, DeserializeError, ValuePointerRef, ErrorKind, deserialize};
+use deserr::serde_json::JsonError;
+use serde_json::json;
+
+#[derive(DeserializeFromValue, Debug, PartialEq, Eq)]
+#[deserr(error = JsonError)]
+struct Search {
+    query: String,
+    limit: usize,
+}
+
+// As we can see, rust is able to infer the error type.
+let data = deserialize::<Search, _, _>(
+    json!({ "query": "doggo", "limit": 1 }),
+)
+.unwrap();
+assert_eq!(data, Search { query: String::from("doggo"), limit: 1 });
+```
+
+It can also be used as a field attribute;
+
+```rust
+use deserr::{DeserializeFromValue, DeserializeError, ValuePointerRef, ErrorKind, deserialize};
+use deserr::serde_json::JsonError;
+use serde_json::json;
+
+// Since the error returned by the `Search` structure needs to implements `MergeWithError<JsonError>`
+// we also need to specify the `error` attribute as a `JsonError`. But as you will see later there are
+// other solutions.
+#[derive(DeserializeFromValue, Debug, PartialEq, Eq)]
+#[deserr(error = JsonError)]
+struct Search<A> {
+    #[deserr(error = JsonError)]
+    query: A,
+    limit: usize,
+}
+```
+
+#### `where_predicate`
+
+Let you add `where` clauses to the `DeserializeFromValue` implementation that deserr will generate.
+
+```rust
+use deserr::{DeserializeFromValue, DeserializeError, MergeWithError, deserialize};
+use deserr::serde_json::JsonError;
+use serde_json::json;
+
+// Here we can constraint the generic `__Deserr_E` type used by deserr to implements `MergeWithError`.
+// Now instead of constraining the final error type it stays generic if it's able to accumulate with
+// with a `JsonError`.
+#[derive(DeserializeFromValue, Debug, PartialEq, Eq)]
+#[deserr(where_predicate = __Deserr_E: MergeWithError<JsonError>, where_predicate = A: DeserializeFromValue<JsonError>)]
+struct Search<A> {
+    #[deserr(error = JsonError)]
+    query: A,
+    limit: usize,
+}
+```
+
+#### `needs_predicate`
+
+Automatically adds `where_predicate = FieldType: DeserializeFromValue<ErrType>` for each field with this attribute.
+
+```rust
+use deserr::{DeserializeFromValue, DeserializeError, MergeWithError, deserialize};
+use deserr::serde_json::JsonError;
+use serde_json::json;
+
+#[derive(DeserializeFromValue, Debug, PartialEq, Eq)]
+struct Search<A> {
+    #[deserr(needs_predicate)]
+    query: A,
+    limit: usize,
+}
+```
+
+Is strictly equivalent to the following:
+
+```rust
+use deserr::{DeserializeFromValue, DeserializeError, MergeWithError, deserialize};
+use deserr::serde_json::JsonError;
+use serde_json::json;
+
+#[derive(DeserializeFromValue, Debug, PartialEq, Eq)]
+#[deserr(where_predicate = A: DeserializeFromValue<__Deserr_E>)]
+struct Search<A> {
+    query: A,
+    limit: usize,
+}
+```
+
+### Comparison with serde
```

Comment:
Could we add somewhere that we don't have a derive macro for tuple structs as well?